### PR TITLE
update boundary interfaces

### PR DESCRIPTION
**Pull request by gassmoeller**
This continues the work of #671 and extends it to the boundary composition and boundary temperature by removing a leftover parameter and unifying the function names. We need a different approach than for the boundary velocity though, because we a *removing* a parameter from the interface that is not known by the interface, therefore we can not implement a new interface function that by default calls the old (deprecated) one, and we do not want to derive the interface from SimulatorAccess to be able to do so. Therefore in this version the Simulator checks if the new function is implemented and outputs a run-time warning if not. It would be nice if somebody could take a look at `core.cc` and say if this is a thing worth pursuing or if there is a cleaner possibility. We can remove the deprecation code in core.cc after the next release or so.

### CODE REVIEW

**Comment by tjhei** [comment]
You have three tests blow up. Let me know if you want me to take a look what is going on there.

File: include/aspect/boundary_composition/interface.h [hunk 1]
```diff
@@ -88,18 +88,45 @@ namespace aspect
...
          */
         virtual
-        double composition (const GeometryModel::Interface<dim> &geometry_model,
-                            const types::boundary_id             boundary_indicator,
-                            const Point<dim>                    &location,
-                            const unsigned int                   compositional_field) const = 0;
+        double
+        boundary_composition (const types::boundary_id             boundary_indicator,
+                              const Point<dim>                    &position,
+                              const unsigned int                   compositional_field) const;
```
Comment (bangerth) [inline include/aspect/boundary_composition/interface.h]: Why give the function a different name? I'm not opposed, just curious why you don't choose the same name as before.

File: include/aspect/utilities.h [hunk 2]
```diff
@@ -45,6 +45,7 @@ namespace aspect
   namespace Utilities
   {
     using namespace dealii;
+    using namespace dealii::Utilities;
```
Comment (bangerth) [inline include/aspect/utilities.h]: Had we settled on this being a good idea? @tjhei ?

File: source/boundary_composition/interface.cc [hunk 3]
```diff
@@ -47,6 +49,32 @@ namespace aspect
...
+      return boundary_composition(boundary_indicator,position,compositional_field);
+    }
+
+    template <int dim>
+    double
+    Interface<dim>::boundary_composition (const types::boundary_id             /*boundary_indicator*/,
+                                          const Point<dim>                    &/*position*/,
+                                          const unsigned int                   /*compositional_field*/) const
+    {
+      return std::numeric_limits<double>::quiet_NaN();
```
Comment (bangerth) [inline source/boundary_composition/interface.cc]: I think you should simply do `Assert(false, ExcMessage("..."));` here. Nothing good can happen if we end up here.

File: source/boundary_composition/spherical_constant.cc [hunk 4]
```diff
@@ -36,22 +37,21 @@ namespace aspect
...
-      // used and what they mean
-      Assert ((dynamic_cast<const GeometryModel::SphericalShell<dim>*>(&geometry_model) != 0
-               || dynamic_cast<const GeometryModel::Chunk<dim>*>(&geometry_model) != 0),
+      // verify that the geometry is a spherical shell, a chunk, or an
+      // ellipsoidal chunk since only for geometries based on spherical shells
+      // do we know which boundary indicators are used and what they mean
+      const GeometryModel::Interface<dim> *geometry_model = &this->get_geometry_model();
+      Assert ((dynamic_cast<const GeometryModel::SphericalShell<dim>*>(geometry_model) != 0
+               || dynamic_cast<const GeometryModel::Chunk<dim>*>(geometry_model) != 0
+               || dynamic_cast<const GeometryModel::EllipsoidalChunk<dim>*>(geometry_model) != 0),
```
Comment (bangerth) [inline source/boundary_composition/spherical_constant.cc]: This is an unrelated change. If not too cumbersome, put it into a separate patch.

File: source/boundary_temperature/spherical_constant.cc [hunk 5]
```diff
@@ -37,22 +37,20 @@ namespace aspect
...
-      Assert ((dynamic_cast<const GeometryModel::SphericalShell<dim>*>(&geometry_model) != 0
-               || dynamic_cast<const GeometryModel::Chunk<dim>*>(&geometry_model) != 0
-               || dynamic_cast<const GeometryModel::EllipsoidalChunk<dim>*>(&geometry_model) != 0),
+      // verify that the geometry is a spherical shell, a chunk, or an
+      // ellipsoidal chunk since only for geometries based on spherical shells
+      // do we know which boundary indicators are used and what they mean
+      const GeometryModel::Interface<dim> *geometry_model = &this->get_geometry_model();
+      Assert ((dynamic_cast<const GeometryModel::SphericalShell<dim>*>(geometry_model) != 0
+               || dynamic_cast<const GeometryModel::Chunk<dim>*>(geometry_model) != 0
+               || dynamic_cast<const GeometryModel::EllipsoidalChunk<dim>*>(geometry_model) != 0),
```
Comment (bangerth) [inline source/boundary_temperature/spherical_constant.cc]: same here

File: source/simulator/core.cc [hunk 6]
```diff
@@ -849,17 +851,46 @@ namespace aspect
...
+          if (std::isnan(boundary_temperature->boundary_temperature(*p,geometry_model->representative_point(0.0))))
+            {
+              if (timestep_number == 0)
+                pcout << "WARNING: Your selected boundary temperature plugin uses the deprecated "
+                      << "interface function 'temperature(geometry_model,boundary_indicator,position)'. "
+                      << "This function will be removed in a future ASPECT "
+                      << "release. Please update your plugin NOW to the new interface "
+                      << "'boundary_temperature(boundary_indicator,position) "
+                      << "instead, and derive it from SimulatorAccess, if you need access to the "
+                      << "geometry model." << std::endl;
```
Comment (bangerth) [inline source/simulator/core.cc]: I'd just abort here. But in fact, you could already do that if you happen to end up in that function as mentioned in my comment there.

Comment (bangerth) [inline source/simulator/core.cc]: It's almost midnight, and I don't think I completely understand how this is supposed to work. Could you outline the idea you're following, maybe with a couple of lines of pseudo-code?

File: include/aspect/boundary_composition/interface.h (hunk 1 above)
Comment (gassmoeller) [inline include/aspect/boundary_composition/interface.h]: This was shortly discussed in #671. The new name synchronizes the names of all different boundary condition plugins. Plus: Considering the discussion we had in #179, it would also synchronize the names of the functions with the (future) names of the plugin sections and folder structure (as soon as I can motivate myself to start implementing #179, which will include a lot of work, and also some incompatibilities).

File: include/aspect/utilities.h (hunk 2 above)
Comment (gassmoeller) [inline include/aspect/utilities.h]: I think @tjhei opposed and I can understand his argument. `dealii::Utilities` contains different utility functions than `aspect::Utilities`. My point was mostly that they are both *utility* functions, and I do not want to remember, which function was in which namespace and would like to avoid the need to specify `aspect::Utilities` vs `dealii::Utilities` every time I use a utility function. I mostly included this for laziness, so if you think we should stick with the old way, I can remove it again.

File: source/boundary_composition/interface.cc (hunk 3 above)
Comment (gassmoeller) [inline source/boundary_composition/interface.cc]: You are right, nothing good can happen if this function is used to calculate something. But for reasons I will discuss below I need a way to check if it is implemented in a particular plugin, without aborting the computation in case it is not. That is why I decided for a default implementation that returns something specifically useless that I can test.

File: source/boundary_composition/spherical_constant.cc (hunk 4 above)
Comment (gassmoeller) [inline source/boundary_composition/spherical_constant.cc]: Ok, I will open another PR.

File: source/simulator/core.cc (hunk 6 above)
Comment (gassmoeller) [inline source/simulator/core.cc]: Sure, it is complicated. 

The problem here is that I want to deprecate the old interface function, to let users know they should update their plugins to the new interface. Usually we do this by deprecating the old function, and implementing a default for the new function that calls the old function, which will throw an exception when both are not implemented in a plugin. In this case however, the new function does not have access to one of the parameters of the old function (the geometry model). So I could only think of one way to do this without breaking every user written plugin: The Simulator needs to check if the new function is implemented, and if it is not, it will output a run-time warning and calls the old function. Therefore I need a default implementation of the new interface function, which returns something that is obviously not the result of a user plugin like a quiet_nan. Using this I can decide within Simulator whether to use the old or new interface. 

So the short answer is: I do not abort, because I want to output a warning for the user, and not throw an error to keep their plugins working the old way for now. It is not pretty to have this additional code in Simulator, but I could not think of another way, and we can remove it after the next release, when users had the chance to update their plugins.

File: include/aspect/boundary_composition/interface.h (hunk 1 above)
Comment (bangerth) [inline include/aspect/boundary_composition/interface.h]: Ah, great. Thanks for the explanation!

File: source/boundary_composition/spherical_constant.cc (hunk 4 above)
Comment (gassmoeller) [inline source/boundary_composition/spherical_constant.cc]: Actually it is more work than I expected, because the new line 47 and the changes in line 48,49 are related to this PR, only line 50 and the update of the documentation is unrelated. If you feel strongly about it I can remove it and open a new PR when this one is merged, but I do not think it is worth the additional work.

Comment (bangerth) [inline source/boundary_composition/spherical_constant.cc]: Let's not worry about it then.

I figure you didn't need this change but saw that it needed to be fixed. In similar cases, I usually just leave the code alone but open an issue on github to remind myself that I need to come back to it.

File: include/aspect/utilities.h (hunk 2 above)
Comment (bangerth) [inline include/aspect/utilities.h]: I'll leave it to @tjhei to make that call.

File: source/simulator/core.cc (hunk 6 above)
Comment (bangerth) [inline source/simulator/core.cc]: I see. So let me rephrase your approach: you're looking for an *exceptional return path* that returns a different kind of value. My choice of words suggests already that maybe throwing (and catching) an *exception* may be the right approach ;-) This avoids the magic value and the `if-else` in the simulator (at the cost of a `try-catch`).

But why not this:
* In the base class, implement the old function and let it call the new function interface. You have all necessary parameters.
* In the base class, implement the new function and let it error out with an `Assert(false)`.
* If a class implements the old interface, this overrides the implementation in the base class and it will work (though no warning message).
* If a class implements the new interface, everything works just fine.
* In the `Simulator`, no change is necessary for now.
The only difference is that you don't get the warning. I think that's ok -- people should read the changelog.

In the end, either approach is fine with me.

**Comment by gassmoeller** [comment]
That would work, I am just not sure if the changelog entry will be enough. I am afraid that nobody will notice and we will end up with a situation like the material model (although there we have some use for a simple interface). If you are fine that we still remove the old interface sometime after the next release (or later), I can rework this to the try-catch structure.

**Comment by bangerth** [comment]
Sure, go for it.

File: include/aspect/utilities.h (hunk 2 above)
Comment (tjhei) [inline include/aspect/utilities.h]: I am fine with it this way.

**Comment by gassmoeller** [comment]
I think I have fixed all issues and comments, but the tester complains about something when uploading the results. @tjhei, could you take a look at the tester results, when you have a minute?

**Comment by bangerth** [comment]
The tests actually succeeded, but the upload failed (because the cdash site died a couple of days ago).
